lock-impl/src/test/java/com/palantir/lock/impl/LegacyTimelockServiceTest.java

```diff
@@ -0,0 +1,203 @@
+
+    private static final long FRESH_TIMESTAMP = 5L;
+
+    private static final LockRefreshTokenV2Adapter LOCK_TOKEN_V2 = randomLockToken();
+    private static final LockRefreshToken LOCK_REFRESH_TOKEN = LOCK_TOKEN_V2.getToken();
+
+    private static final LockDescriptor LOCK_A = StringLockDescriptor.of("a");
+    private static final LockDescriptor LOCK_B = StringLockDescriptor.of("b");
+
+    private final TimestampService timestampService = mock(TimestampService.class);
+    private final RemoteLockService lockService = mock(RemoteLockService.class);
+
+    private final TimelockService timelock = new LegacyTimelockService(timestampService, lockService, LOCK_CLIENT);
+
+    @Before
+    public void before() {
+        when(timestampService.getFreshTimestamp()).thenReturn(FRESH_TIMESTAMP);
+    }
+
+    @Test
+    public void freshTimestampDelegatesToTimestampService() {
+        assertEquals(FRESH_TIMESTAMP, timelock.getFreshTimestamp());
+    }
+
+    @Test
+    public void freshTimestampsDelegatesToTimestampService() {
+        int numTimestamps = 10;
+        TimestampRange range = TimestampRange.createInclusiveRange(21L, 30L);
+        when(timestampService.getFreshTimestamps(10)).thenReturn(range);
+
+        assertEquals(range, timelock.getFreshTimestamps(numTimestamps));
+    }
+
+    @Test
+    public void lockImmutableTimestampLocksFreshTimestamp() throws InterruptedException {
+        long immutableTs = 3L;
+
+        mockImmutableTsLockResponse();
+        mockMinLockedInVersionIdResponse(immutableTs);
+
+        LockImmutableTimestampResponse expectedResponse = LockImmutableTimestampResponse.of(immutableTs, LOCK_TOKEN_V2);
+        assertEquals(expectedResponse, timelock.lockImmutableTimestamp(LockImmutableTimestampRequest.create()));
+    }
+
+    @Test
+    public void getImmutableTimestampDelegatesInProperOrder() throws InterruptedException {
+        long immutableTs = 3L;
+
+        InOrder inOrder = Mockito.inOrder(timestampService, lockService);
+
+        mockMinLockedInVersionIdResponse(immutableTs);
+
+        assertEquals(immutableTs, timelock.getImmutableTimestamp());
+        inOrder.verify(timestampService).getFreshTimestamp();
+        inOrder.verify(lockService).getMinLockedInVersionId(LOCK_CLIENT.getClientId());
+    }
+
+    @Test
+    public void getImmutableTimestampReturnsFreshTimestampIfMinLockedInVersionIsNull() throws InterruptedException {
+        mockMinLockedInVersionIdResponse(null);
+
+        assertEquals(FRESH_TIMESTAMP, timelock.getImmutableTimestamp());
+    }
+
+    @Test
+    public void lockDelegatesToLockService() throws InterruptedException {
+        LockRequest legacyRequest = LockRequest.builder(buildLockMap(LockMode.WRITE)).build();
+
+        when(lockService.lock(LockClient.ANONYMOUS.getClientId(), legacyRequest)).thenReturn(LOCK_REFRESH_TOKEN);
+
+        assertEquals(LOCK_REFRESH_TOKEN, timelock.lock(LockRequestV2.of(ImmutableSet.of(LOCK_A, LOCK_B))));
```

Comment:
`assertEquals(LOCK_TOKEN_V2, timelock.lock(LockRequestV2.of(ImmutableSet.of(LOCK_A, LOCK_B))));`